crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+{
+    OSSL_HPKE_SUITE chosen;
+    int crv = 0;
+    int erv = 0;
+    size_t plen = 0;
+    uint16_t kem_ind = 0;
+
+    if (pub == NULL || !pub_len
+        || ciphertext == NULL || !ciphertext_len || suite == NULL)
+        return 0;
+    if (suite_in == NULL) {
+        /* choose a random suite */
+        crv = hpke_random_suite(libctx, propq, &chosen);
+        if (crv != 1)
+            return crv;
+    } else {
+        chosen = *suite_in;
+    }
+    kem_ind = kem_iana2index(chosen.kem_id);
+    if (kem_ind == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if ((crv = hpke_suite_check(chosen)) != 1)
+        return 0;
+    *suite = chosen;
+    /* publen */
+    plen = hpke_kem_tab[kem_ind].Npk;
+    if (plen > *pub_len)
+        return 0;
+    if (RAND_bytes_ex(libctx, pub, plen, OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    *pub_len = plen;
+    if (RAND_bytes_ex(libctx, ciphertext, ciphertext_len,
+                      OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    return 1;
+err:
+    return erv;
+}
+
+/*
+ * @brief string matching for suites
+ */
+#if defined(_WIN32)
+# define HPKE_MSMATCH(inp, known) \
+    (strlen(inp) == strlen(known) && !_stricmp(inp, known))
+#else
+# define HPKE_MSMATCH(inp, known) \
+    (strlen(inp) == strlen(known) && !strcasecmp(inp, known))
+#endif
+
+/*
+ * @brief map a string to a HPKE suite
+ *
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
+{
+    uint16_t kem = 0, kdf = 0, aead = 0;
+    char *st = NULL;
+    char *instrcp = NULL;
+    size_t inplen = 0;
+    int labels = 0;
+
+    if (suitestr == NULL || suite == NULL)
+        return 0;
+    /* See if it contains a mix of our strings and numbers  */
+    inplen = OPENSSL_strnlen(suitestr, OSSL_HPKE_MAX_SUITESTR);
+    if (inplen >= OSSL_HPKE_MAX_SUITESTR)
+        return 0;
+    instrcp = OPENSSL_strndup(suitestr, inplen);
+    st = strtok(instrcp, ",");
+    if (st == NULL) {
+        OPENSSL_free(instrcp);
+        return 0;
+    }
+    while (st != NULL && ++labels <= 3) {
+        /* check if string is known or number and if so handle appropriately */
+        if (kem == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P256)) {
+                kem = OSSL_HPKE_KEM_ID_P256;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P384)) {
+                kem = OSSL_HPKE_KEM_ID_P384;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P521)) {
+                kem = OSSL_HPKE_KEM_ID_P521;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X25519)) {
+                kem = OSSL_HPKE_KEM_ID_X25519;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X448)) {
+                kem = OSSL_HPKE_KEM_ID_X448;
+            }
```

Comment:
I would prefer to see this kind of thing being table driven.. 
e.g idmap[] = { 
    {OSSL_HPKE_KEM_ID_P256, { OSSL_HPKE_KEMSTR_P256, "0x10", "16"} },
}

Reply:
did that - I'm not sure the end result is better (*) tbh, but if that's what's preferred....

(*) I'd argue that the previous very-verbose version was much easier to maintain whereas this terse version will likely be harder to modify in all but one way